### PR TITLE
ConditionSet

### PR DESCRIPTION
**Pull request by ranjithkumar007**
ping @nishnik.

### CODE REVIEW

File: symengine/sets.cpp [hunk 1]
```diff
@@ -94,7 +94,7 @@ RCP<const Set> Interval::close() const
 
 RCP<const Boolean> Interval::contains(const RCP<const Basic> &a) const
 {
-    if (not is_a_Number(*a))
+    if (not is_a_Number(*a) and not is_a<Constant>(*a))
```
Comment (isuruf) [inline symengine/sets.cpp]: Does this work?

File: symengine/sets.h [hunk 2]
```diff
@@ -249,6 +249,46 @@ class Complement : public Set
     }
 };
 
+class ConditionSet : public Set
+{
+private:
+    vec_sym syms_;
+    RCP<const Set> base_;
```
Comment (isuruf) [inline symengine/sets.h]: Why can't this be in condition?

File: symengine/sets.cpp (hunk 1 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: no. I shall remove this.

**Comment by ranjithkumar007** [comment]
@isuruf ,
Should `And(a,b)` evaluate to `contains(x,[1,2])` where `a=contains(x,[1,5])` and `b=contains(x,[1,2])` or remain as unevaluated?

**Comment by ranjithkumar007** [comment]
ping @isuruf

**Comment by Sumith1896** [comment]
Ah this is interesting :D

File: symengine/logic.h [hunk 3]
```diff
@@ -65,6 +65,8 @@ class Contains : public Boolean
     RCP<const Set> get_set() const;
     virtual vec_basic get_args() const;
     virtual bool __eq__(const Basic &o) const;
+    virtual RCP<const Basic> create(const RCP<const Basic> &lhs,
```
Comment (isuruf) [inline symengine/logic.h]: Is `virtual` needed?

Comment (ranjithkumar007) [inline symengine/logic.h]: ah. nope
Thanks

**Comment by ranjithkumar007** [comment]
ping @Sumith1896 @nishnik

File: symengine/tests/logic/test_logic.cpp [hunk 4]
```diff
@@ -137,7 +137,7 @@ TEST_CASE("And, Or : Basic", "[basic]")
     auto int1 = interval(integer(1), integer(2), false, false);
     auto int2 = interval(integer(1), integer(5), false, false);
     auto c1 = contains(x, int1);
-    auto c2 = contains(x, int2);
+    auto c2 = contains(symbol("y"), int2);
```
Comment (isuruf) [inline symengine/tests/logic/test_logic.cpp]: why this change?

File: symengine/tests/logic/test_logic.cpp [hunk 5]
```diff
@@ -237,7 +237,7 @@ TEST_CASE("Not : Basic", "[basic]")
...
```
Comment (isuruf) [inline symengine/tests/logic/test_logic.cpp]: Why this change?

File: symengine/sets.cpp [hunk 6]
```diff
@@ -620,6 +632,123 @@ RCP<const Set> Complement::set_complement(const RCP<const Set> &o) const
...
+    return make_rcp<const Complement>(o, rcp_from_this_cast<const Set>());
+}
+
+RCP<const Boolean> ConditionSet::contains(const RCP<const Basic> &o) const
+{
+    if (is_a<FiniteSet>(*o)) {
+        const FiniteSet &fs = down_cast<const FiniteSet &>(*o);
+        auto container = fs.get_container();
+        if (syms_.size() != container.size()) {
+            throw std::runtime_error("size of symbols didn't match");
```
Comment (isuruf) [inline symengine/sets.cpp]: Why does this have to match?

File: symengine/sets.h [hunk 7]
```diff
@@ -249,6 +249,39 @@ class Complement : public Set
     }
 };
 
+class ConditionSet : public Set
+{
+private:
+    vec_sym syms_;
```
Comment (isuruf) [inline symengine/sets.h]: Why is this a vector? SymPy only has one symbol?

Comment (ranjithkumar007) [inline symengine/sets.h]:
```
>>> ConditionSet({x,y},x+y>4,S.Reals)
{{x, y} | {x, y} ∊ ℝ ∧ x + y > 4}
```
SymPy supports multiple symbols.

File: symengine/tests/logic/test_logic.cpp (hunk 4 above)
Comment (ranjithkumar007) [inline symengine/tests/logic/test_logic.cpp]: I will revert this. I earlier changed this because initially I simplified `contains`.

File: symengine/sets.h (hunk 7 above)
Comment (isuruf) [inline symengine/sets.h]: Is this intended or not? I don't see any documentation with multiple symbols.

Comment (ranjithkumar007) [inline symengine/sets.h]: https://github.com/sympy/sympy/blob/master/sympy/solvers/tests/test_solveset.py#L1383
This is the only point I could find where SymPy actually uses multiple symbols for ConditionSet.
Should we make it only for a single variable then?

Comment (isuruf) [inline symengine/sets.h]: Yes, you can make it a single variable and you can use a FiniteSet (which is a single entity) if there is a need

Comment (ranjithkumar007) [inline symengine/sets.h]: for allowing it to be `FiniteSet`, `RCP<const Basic>` should be the data member right instead of `RCP<const Symbol>`?
We can currently define it only if the input is of type Symbol.

Comment (isuruf) [inline symengine/sets.h]: Yes. `RCP<const Basic>` is good enough.

**Comment by ranjithkumar007** [comment]
thanks @isuruf-bot  :D

File: symengine/sets.cpp [hunk 8]
```diff
@@ -620,6 +632,107 @@ RCP<const Set> Complement::set_complement(const RCP<const Set> &o) const
...
 
+ConditionSet::ConditionSet(const RCP<const Basic> sym,
+                           RCP<const Boolean> condition)
+    : sym(sym), condition_(condition)
+{
+    SYMENGINE_ASSIGN_TYPEID()
+    SYMENGINE_ASSERT(ConditionSet::is_canonical(sym, condition))
+}
+
+bool ConditionSet::is_canonical(const RCP<const Basic> sym,
```
Comment (srajangarg) [inline symengine/sets.cpp]: If `condition` does not involve symbol at all `sym`, it should be the `UniversalSet` and return `false`?

Comment (srajangarg) [inline symengine/sets.cpp]: What I mean is something like `{x | y > 0}`

Comment (ranjithkumar007) [inline symengine/sets.cpp]: if we return `UniversalSet`, then it would guarantee that for any x, condition `y>0`  is true, which might not be true.

Comment (srajangarg) [inline symengine/sets.cpp]: I'm not versed with the theory involved, it should either be `Universal` or `Empty`. @nishnik ?

Comment (ranjithkumar007) [inline symengine/sets.cpp]: ping @nishnik

Comment (nishnik) [inline symengine/sets.cpp]: I think it should remain un-evaluated as SymPy does.
Because if in any case, `x` gets dependent on `y`, we can't be sure that it is universal set or empty set (it can be an interval as well, in your example put `y=x+5`)
And if we consider that `x` and `y` are always going to be independent, which can easily be discarded by using three inter-dependent variables in reasoning.

Comment (ranjithkumar007) [inline symengine/sets.cpp]: Agreed @nishnik.

**Comment by ranjithkumar007** [comment]
can you have a look @srajangarg @nishnik ?

**Comment by srajangarg** [comment]
LGTM @nishnik  any comments? or should I merge?

File: symengine/subs.h [hunk 9]
```diff
@@ -144,6 +144,25 @@ class SubsVisitor : public BaseVisitor<SubsVisitor>
         result_ = x.create(v);
     }
 
+    void bvisit(const Contains &x)
+    {
+        RCP<const Basic> a = apply(x.get_expr());
+        RCP<const Set> b = rcp_static_cast<const Set>(apply(x.get_set()));
```
Comment (isuruf) [inline symengine/subs.h]: How can you be sure that the result is a `Set`

File: symengine/subs.h [hunk 10]
```diff
@@ -144,6 +144,25 @@ class SubsVisitor : public BaseVisitor<SubsVisitor>
...
+            result_ = x.rcp_from_this();
+        else
+            result_ = x.create(a, b);
+    }
+
+    void bvisit(const And &x)
+    {
+        set_boolean v;
+        for (const auto &elem : x.get_container()) {
+            v.insert(rcp_static_cast<const Boolean>(apply(elem)));
```
Comment (isuruf) [inline symengine/subs.h]: Same here

File: symengine/tests/basic/test_sets.cpp [hunk 11]
```diff
@@ -480,7 +493,7 @@ TEST_CASE("Complement : Basic", "[basic]")
     REQUIRE(eq(*r4.get_universe(), *i1));
 
     REQUIRE(r1->get_args().size() == 2);
-    REQUIRE(eq(*r1->contains(one), *boolTrue));
+    REQUIRE(is_a<Not>(*r1->contains(one)));
```
Comment (isuruf) [inline symengine/tests/basic/test_sets.cpp]: Why this change?

File: symengine/tests/basic/test_sets.cpp [hunk 12]
```diff
@@ -494,8 +507,8 @@ TEST_CASE("Complement : Basic", "[basic]")
 
     CHECK_THROWS_AS(r2->set_intersection(finiteset({symbol("x")})),
                     std::runtime_error);
-    r1 = r2->set_intersection(finiteset({zero, integer(2)}));
-    REQUIRE(eq(*r1, *finiteset({zero, integer(2)})));
+    CHECK_THROWS_AS(r2->set_intersection(finiteset({zero, integer(2)})),
+                    std::runtime_error);
```
Comment (isuruf) [inline symengine/tests/basic/test_sets.cpp]: Why this change?

File: symengine/sets.cpp [hunk 13]
```diff
@@ -804,4 +917,59 @@ RCP<const Set> set_complement(const RCP<const Set> &universe,
     return container->set_complement(universe);
 }
 
+RCP<const Set> conditionset(const RCP<const Basic> &sym,
+                            const RCP<const Boolean> &condition)
+{
+    if (ConditionSet::is_canonical(sym, condition)) {
```
Comment (isuruf) [inline symengine/sets.cpp]: Why this? Everything is checked twice because of this

File: symengine/sets.cpp [hunk 14]
```diff
@@ -804,4 +917,59 @@ RCP<const Set> set_complement(const RCP<const Set> &universe,
...
+    if (is_a<And>(*condition)) {
+        // Simplify if we have a single symbol.
+        const And &aset = down_cast<const And &>(*condition);
+        auto &container = aset.get_container();
+        set_basic present, others;
+        for (auto it = container.begin(); it != container.end(); it++) {
+            if (is_a<Contains>(**it)
+                and eq(*down_cast<const Contains &>(**it).get_expr(), *sym)
+                and is_a<FiniteSet>(
+                        *down_cast<const Contains &>(**it).get_set())) {
```
Comment (isuruf) [inline symengine/sets.cpp]: Can you comment in the code explaining what you are doing here?

File: symengine/subs.h (hunk 9 above)
Comment (ranjithkumar007) [inline symengine/subs.h]: `x.get_set()` always returns the object of type `Set`, then doesn't that guarantee that output is a `Set`?

File: symengine/tests/basic/test_sets.cpp (hunk 11 above)
Comment (ranjithkumar007) [inline symengine/tests/basic/test_sets.cpp]: at this point, `r1` stores `(-oo, oo) \ {y} `. But here we are not sure if `finiteset({y})` contains `one` or not.
So, this should return `not(contains(1,{y}))`. Earlier this bug was overshadowed by the bug in FintieSet `contains`, which I corrected now.

File: symengine/tests/basic/test_sets.cpp (hunk 12 above)
Comment (ranjithkumar007) [inline symengine/tests/basic/test_sets.cpp]: similar argument as ^above here.
`r2` contains `(-oo, oo) \ {y}`. so we can't simplify this unless we have info of `y`. So, this should return an object of type `Intersection` class which is currently not implemented. So, `runtime_error`.

File: symengine/sets.cpp (hunk 13 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: Thanks. Will be updated.

File: symengine/subs.h (hunk 9 above)
Comment (ranjithkumar007) [inline symengine/subs.h]: @isuruf , is there any alternate way of doing this then?

File: symengine/logic.cpp [hunk 15]
```diff
@@ -105,9 +105,9 @@ int Contains::compare(const Basic &o) const
 }
 
 RCP<const Basic> Contains::create(const RCP<const Basic> &lhs,
-                                  const RCP<const Set> &rhs) const
+                                  const RCP<const Basic> &rhs) const
 {
-    return contains(lhs, rhs);
+    return contains(lhs, rcp_static_cast<const Set>(rhs));
```
Comment (isuruf) [inline symengine/logic.cpp]: Why? You should never use `rcp_static_cast` if you are unsure of the type

File: symengine/logic.cpp [hunk 16]
```diff
@@ -215,9 +215,13 @@ const set_boolean &And::get_container() const
     return container_;
 }
 
-RCP<const Basic> And::create(const set_boolean &a) const
+RCP<const Basic> And::create(const vec_basic &a) const
```
Comment (isuruf) [inline symengine/logic.cpp]: Why?

File: symengine/sets.cpp [hunk 17]
```diff
@@ -804,4 +917,70 @@ RCP<const Set> set_complement(const RCP<const Set> &universe,
...
+                // iterate through container and check for the condition that
+                // defines the domain of sym.
+                // Simplify if that set is a FiniteSet.
```
Comment (isuruf) [inline symengine/sets.cpp]: These types of simplification logic does not belong here. If `And` can be simplified to a `Contains` with a `FiniteSet`, then that should happen in `And`.

File: symengine/logic.cpp (hunk 15 above)
Comment (ranjithkumar007) [inline symengine/logic.cpp]: then how should this be done?

Comment (ranjithkumar007) [inline symengine/logic.cpp]: I mean how can I call `contains` whose second param is a `Set`?

File: symengine/sets.cpp (hunk 17 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: Got it.

File: symengine/logic.cpp (hunk 15 above)
Comment (isuruf) [inline symengine/logic.cpp]: You have to check that the object is a `Set` and throw an error if not.

Comment (ranjithkumar007) [inline symengine/logic.cpp]: Okay ! Thanks

Comment (ranjithkumar007) [inline symengine/logic.cpp]: this check is supposed to be done here in `create` rather than in `subs` right?

Comment (isuruf) [inline symengine/logic.cpp]: No, in `subs`.

**Comment by ranjithkumar007** [comment]
ping @isuruf

File: symengine/sets.cpp [hunk 18]
```diff
@@ -804,4 +911,46 @@ RCP<const Set> set_complement(const RCP<const Set> &universe,
...
+                }
+            } else {
+                newcont.insert(*it);
+            }
+        }
+        if (not temp.empty()) {
+            return SymEngine::set_union(
+                {SymEngine::set_union(temp),
+                 conditionset(sym, logical_and(newcont))});
+        }
```
Comment (isuruf) [inline symengine/sets.cpp]: Can you explain what's the purpose of the lines 923-948?

Comment (ranjithkumar007) [inline symengine/sets.cpp]: say `logical_and({0,1,-3,y}->contains(x),x>0)` is the given condition.
Then this gets simplified to `And({1}->contains(x),{y}->contains(x),x>0)` within `logical_and()`.
Now, here we need to extract this `{1}` from the `And` to return the output of `conditionset()`as `{1}UConditionSet(x,logical_and({y}->contains(x),x>0))`. Basically, it seperates out all such `finitesets` which satisy the given condition for sure.

Comment (isuruf) [inline symengine/sets.cpp]: `logical_and({0,1,-3,y}->contains(x),x>0)` is not equivalent to `And({1}->contains(x),{y}->contains(x),x>0)`

Comment (ranjithkumar007) [inline symengine/sets.cpp]: why ?can't we ignore other terms in the Finite set as they don't satisfy `x>0` condition?

Comment (isuruf) [inline symengine/sets.cpp]: `logical_and({0,1,-3,y}->contains(x),x>0)` is equivalent to `And(Or({1}->contains(x),{y}->contains(x)),x>0)`

Comment (ranjithkumar007) [inline symengine/sets.cpp]: my bad! cool, Now I need to correct that simplification in And() and also add a condition here in `conditionset()`  to consider arguments of form `Or()` also?

Comment (isuruf) [inline symengine/sets.cpp]: I'd prefer if `logical_and({0,1,-3,y}->contains(x),x>0)` was simplified to `logical_and({1,y}->contains(x),x>0)`

Comment (isuruf) [inline symengine/sets.cpp]: Also, don't do simplifications you are not sure are correct. It's better the result be in non-simplified form rather than an incorrect result.

Comment (ranjithkumar007) [inline symengine/sets.cpp]: Okay! Thanks !!

Comment (isuruf) [inline symengine/sets.cpp]: SymPy doesn't do the simplification either, `ConditionSet(x, And(FiniteSet(-1, 0, 2).contains(x), x > 0), S.Reals)`

File: symengine/sets.cpp [hunk 19]
```diff
@@ -620,6 +632,101 @@ RCP<const Set> Complement::set_complement(const RCP<const Set> &o) const
...
+        for (const auto &c : cont) {
+            d[sym] = c;
+            if (not(eq(*condition_->subs(d), *boolTrue)))
+                return boolean(false);
+            d.clear();
+        }
+        return boolean(true);
+    }
+    d[sym] = o;
+    return boolean(eq(*condition_->subs(d), *boolTrue));
```
Comment (isuruf) [inline symengine/sets.cpp]: This is wrong too. `ConditionSet(x, sin(x)<y, S.Reals).contains(5)`

File: symengine/sets.cpp [hunk 20]
```diff
@@ -620,6 +632,101 @@ RCP<const Set> Complement::set_complement(const RCP<const Set> &o) const
...
+
+RCP<const Boolean> ConditionSet::contains(const RCP<const Basic> &o) const
+{
+    map_basic_basic d;
+    if (is_a<FiniteSet>(*o)) {
+        auto &cont = down_cast<const FiniteSet &>(*o).get_container();
+        for (const auto &c : cont) {
+            d[sym] = c;
+            if (not(eq(*condition_->subs(d), *boolTrue)))
+                return boolean(false);
```
Comment (isuruf) [inline symengine/sets.cpp]: Here you are checking the `ConditionSet` contains all of the elements of `o`. This is not what `contains` means. Contains means you check that the element `o` is contained in the set. `o` can be a element only if `sym` is a FiniteSet and this case is already handled in Line 726 and below

File: symengine/sets.cpp (hunk 19 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: I think it should return `condition_->subs(d)`.

File: symengine/sets.cpp (hunk 18 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: should we skip simplifying this?

Comment (ranjithkumar007) [inline symengine/sets.cpp]: @isuruf, actually SymPy does this simplification. when we merged the base set into this condition, complicacy increased. 
```
>>> ConditionSet(x,x>0,FiniteSet(0,1,-3,y))
{1} ∪ {x | x ∊ {y} ∧ x > 0}
```

Comment (isuruf) [inline symengine/sets.cpp]: Okay, then let's keep this

File: symengine/sets.cpp (hunk 19 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: i mean rcp_static_cast<const Boolean>(condition_->subs(d)), after checking if this is of type Boolean.

File: symengine/tests/basic/test_sets.cpp [hunk 21]
```diff
@@ -690,16 +698,22 @@ TEST_CASE("ConditionSet : Basic", "[basic]")
     REQUIRE(eq(*r1->contains(integer(3)), *boolTrue));
     REQUIRE(eq(*r1->contains(integer(-3)), *boolFalse));
     REQUIRE(eq(*r1->contains(integer(2)), *boolFalse));
-    REQUIRE(eq(*r1->contains(finiteset({integer(3)})), *boolTrue));
+    REQUIRE(eq(*r1->contains(finiteset({integer(3)})), *boolFalse));
```
Comment (ranjithkumar007) [inline symengine/tests/basic/test_sets.cpp]: @isuruf, this test case is failing as  `And(9 == {3}**2, Contains({3}, (1, oo)))` doesn't evaluate to `boolFalse`.
Is this a problem with `Interval:: contains()`?

File: symengine/sets.cpp [hunk 22]
```diff
@@ -804,4 +905,48 @@ RCP<const Set> set_complement(const RCP<const Set> &universe,
...
+                    }
+                }
+            } else {
+                newcont.insert(*it);
+            }
+        }
+        if (not present.empty()) {
+            newcont.insert(finiteset(others)->contains(sym));
+            return SymEngine::set_union(
+                {finiteset(present), conditionset(sym, logical_and(newcont))});
```
Comment (isuruf) [inline symengine/sets.cpp]: What's the result of `And(Contains(x, {1, 2, 3}), Eq(y, x))` and `conditionset(x, And(Contains(x, {1, 2, 3}), Eq(y, x)))`?

File: symengine/tests/basic/test_sets.cpp (hunk 21 above)
Comment (isuruf) [inline symengine/tests/basic/test_sets.cpp]: @nishnik might be more familiar.

File: symengine/sets.cpp (hunk 22 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: `And(x == y, Contains(x, {1, 2, 3}))` and `{1, 2, 3}` respectively.

Comment (isuruf) [inline symengine/sets.cpp]: Okay, we need to fix the the second result then.

File: symengine/tests/basic/test_sets.cpp (hunk 21 above)
Comment (ranjithkumar007) [inline symengine/tests/basic/test_sets.cpp]: I think that the problem is with `Interval:: contains`
```
diff --git a/symengine/logic.cpp b/symengine/logic.cpp
index 979ff74..24820e2 100644
--- a/symengine/logic.cpp
+++ b/symengine/logic.cpp
 RCP<const Boolean> contains(const RCP<const Basic> &expr,
                             const RCP<const Set> &set)
 {
-    if (is_a_Number(*expr)) {
-        return set->contains(expr);
-    } else {
-        return make_rcp<Contains>(expr, set);
-    }
+    return set->contains(expr);
 }
 
diff --git a/symengine/sets.cpp b/symengine/sets.cpp
--- a/symengine/sets.cpp
+++ b/symengine/sets.cpp
 
 RCP<const Boolean> Interval::contains(const RCP<const Basic> &a) const
 {
-    if (not is_a_Number(*a))
-        return make_rcp<Contains>(a, rcp_from_this_cast<const Set>());
+    if (not is_a_Number(*a)) {
+        if (is_a<Symbol>(*a) or is_a<Constant>(*a))
+            return make_rcp<Contains>(a, rcp_from_this_cast<const Set>());
+        return boolean(false);
+    }

```
^Above patch should fix this problem. thoughts @isuruf and @nishnik?

Comment (isuruf) [inline symengine/tests/basic/test_sets.cpp]: With that patch `interval(1, 2)->contains(y**2)` returns false.

Comment (ranjithkumar007) [inline symengine/tests/basic/test_sets.cpp]: how to make `Contains({3}, (1, oo))`  return false then ?

Comment (ranjithkumar007) [inline symengine/tests/basic/test_sets.cpp]: SymPy returns false if `inp` is not a real number.

Comment (isuruf) [inline symengine/tests/basic/test_sets.cpp]: You can do as sympy does and return false for any `Set`s

File: symengine/sets.cpp [hunk 23]
```diff
@@ -935,7 +925,27 @@ RCP<const Set> conditionset(const RCP<const Basic> &sym,
                     if (not(is_a_Number(*elem) or is_a<Constant>(*elem))) {
                         others.insert(elem);
                     } else {
-                        present.insert(elem);
+                        // logical_and() doesn't guarantee that if element of a
```
Comment (ranjithkumar007) [inline symengine/sets.cpp]: @isuruf , can you have a look here?
I added a test you suggested for this below.

**Comment by ranjithkumar007** [comment]
ping @isuruf.

File: symengine/sets.cpp [hunk 24]
```diff
@@ -620,6 +637,85 @@ RCP<const Set> Complement::set_complement(const RCP<const Set> &o) const
...
+    return make_rcp<const Complement>(o, rcp_from_this_cast<const Set>());
+}
+
+RCP<const Boolean> ConditionSet::contains(const RCP<const Basic> &o) const
+{
+    map_basic_basic d;
+    d[sym] = o;
+    auto cond = condition_->subs(d);
+    if (not is_a_Boolean(*cond)) {
+        throw std::runtime_error("expected an object of type Boolean");
```
Comment (isuruf) [inline symengine/sets.cpp]: Throw `SymEngineException`. Don't use `std::runtime_error`

**Comment by ranjithkumar007** [comment]
updated @isuruf

**Comment by ranjithkumar007** [comment]
Thanks a lot for the thorough review. :)
